### ISSUE TITLE
CUDA RNG is not reproducible across GPU models above a launch-geometry threshold

### ISSUE THREAD

**Issue by 59thProfile** [pytorch/pytorch#194676]
### 🐛 Describe the bug

[A205_ISSUE_READY.md](https://github.com/user-attachments/files/31395209/A205_ISSUE_READY.md)

### Summary

Seeding with `torch.manual_seed(s)` + `torch.cuda.manual_seed_all(s)` and drawing from a CUDA
sampler gives **different values on different NVIDIA GPU models** once the tensor exceeds a
threshold determined by the **launch geometry**, not by the seed.

The transition is **exactly one element wide**. Measured on Tesla T4 (40 SM) and NVIDIA L4 (58 SM),
torch 2.11.0+cu128:

| `numel` | sampler entry points agreeing across T4 and L4 |
|---|---|
| **40,960** | **8 / 8** |
| **40,961** | **0 / 8** |
| 163,840 | 0 / 8 |
| 163,841 | 0 / 8 |

40,960 is exactly `40 SM × (1024 threads/SM ÷ 256) × 256` — the T4's thread capacity for these
kernels. The L4's is 89,088.

**`torch.use_deterministic_algorithms(True)` does not cover this.** On both devices the flagged and
unflagged hashes are identical, and both still differ across devices:

```
T4   normal_(40961) flagged=1e6b6814d99649e0 unflagged=1e6b6814d99649e0
L4   normal_(40961) flagged=944733fd5370b130 unflagged=944733fd5370b130
```

---

### Reproduce

```bash
python3 repro_aten_offset.py     # on GPU model A
python3 repro_aten_offset.py     # on GPU model B
diff                             # compare the printed hashes
```

Requires only `pip install torch`. No model, no dataset, no other dependency.

### Measured output

```
device=Tesla T4  sm=40 thr/sm=1024 threads_at_cap=40960 torch=2.11.0+cu128
  normal_       40960:b6b67ebb834a27e4  40961:1e6b6814d99649e0
  uniform_      40960:eb3808fdeca10112  40961:33a5cbbd8c543953
  bernoulli_    40960:d6f62c8682c3cc7d  40961:ce6795406abca5c4
  exponential_  40960:ecfeaacdc96e0657  40961:c6208df814fea198
  randn         40960:b6b67ebb834a27e4  40961:1e6b6814d99649e0
  rand          40960:eb3808fdeca10112  40961:33a5cbbd8c543953
  randint       40960:fce44208b62d3090  40961:5365e46736ba7f05
  randperm      40960:7f3db55006fca09e  40961:e9768dd4be9c3a31

device=NVIDIA L4 sm=58 thr/sm=1536 threads_at_cap=89088 torch=2.11.0+cu128
  normal_       40960:b6b67ebb834a27e4  40961:944733fd5370b130     <-- 40960 SAME, 40961 DIFFERS
  uniform_      40960:eb3808fdeca10112  40961:c8ab3cbd8be8a735
  bernoulli_    40960:d6f62c8682c3cc7d  40961:8b50cdad6d547631
  exponential_  40960:ecfeaacdc96e0657  40961:7482c2ffea6aab05
  randn         40960:b6b67ebb834a27e4  40961:944733fd5370b130
  rand          40960:eb3808fdeca10112  40961:c8ab3cbd8be8a735
  randint       40960:fce44208b62d3090  40961:282568f3a853ae9b
  randperm      40960:7f3db55006fca09e  40961:41156a77262e45d8
```

---

### Mechanism

Two facts compose:

1. **`calc_execution_policy` clamps the grid by SM count.**
   ```
   grid.x = min( SMs * (maxThreadsPerSM / block_size),
                 ceil(numel / (block_size * VEC)) )
   ```
   When `numel` exceeds the device's thread capacity, the grid saturates and each thread handles
   **more than one element**, with a stride equal to the total thread count.

2. **`curand_init(seed, idx, offset, &state)` seeds Philox with the *thread index*.**
   The random stream is therefore attached to a **thread**, not to an **element**.

Consequently the **element → thread** mapping is a function of the device's SM count, and so are
the values. A device with fewer threads wraps; a device with more does not; they disagree.

The seed is respected. The *assignment of seeded streams to elements* is not portable.

### Threshold

```
T_device = SMs × ⌊maxThreadsPerSM / 256⌋ × 256
```

Two devices agree iff `ceil(numel / VEC) ≤ min(T_A, T_B)`.

| device | SMs | thr/SM | T |
|---|---|---|---|
| Tesla T4 (sm_75) | 40 | 1024 | **40,960** |
| NVIDIA L4 (sm_89) | 58 | 1536 | **89,088** |

**`VEC` shifts the threshold by a whole family.** Kernels using a contiguous 4-wide vector per
thread (the dropout family) flip at **4×** the distribution-kernel threshold — 163,841 rather than
40,961 for this device pair.

**The unsafe set is not an interval.** `VEC` is the widest aligned access that *divides* `numel`, so
whether a given size is safe depends on its divisibility, producing a comb rather than a cutoff.
The reproducer's four sizes straddle both family thresholds; a full comb characterisation is
outside this report.

---

### The mechanism, and one thing this reproducer does NOT show

**The cross-device effect is the within-call element→thread remapping, and that is sufficient on its
own.** §3 of the reproducer isolates it: re-seeding immediately before each call removes any
possibility of a carried offset, and the values still differ across devices.

```
T4   (a) remap-only, re-seeded each call: 1ad635fce55bcc18 1ad635fce55bcc18 equal=True
L4   (a) remap-only, re-seeded each call: b83c54926ac8f147 b83c54926ac8f147 equal=True
```

Equal on one device across repeats; **different between the two devices**. That is the finding.

### Philox offset carry-over is NOT a second cross-device mechanism — corrected before filing

An earlier draft of this report presented offset carry-over as an independent mechanism and used it
to argue severity. **Our own witness refutes that, and the refuting lines are printed above in the
very output we are pasting:**

```
T4   (b) small draw AFTER a big one: 23b65da237a20781   the same draw ALONE: 7274c7c74a8a05fb
L4   (b) small draw AFTER a big one: 23b65da237a20781   the same draw ALONE: 7274c7c74a8a05fb
```

**Byte-for-byte identical across the two devices.** The offset advance was the *same* on two
different launch geometries and produced **no cross-device divergence whatsoever**. ATen's
`counter_offset = ((numel−1)/(block·grid·unroll)+1)·unroll` evaluates to **4** on both devices at
`numel = 40961`: the *grid* differs — that is mechanism (a), and it is real — the *offset* does not.

What (b) does show is that a preceding draw advances a counter-based generator. That is trivially
true at any size, has **no threshold sensitivity**, and this reproducer contains **no
below-threshold control** for it. It is therefore not evidence of anything device-specific and is
withdrawn from this report.

### Severity, re-argued on (a) alone

`[ARGUED]` — this is arithmetic about the threshold, not a measurement in this reproducer.
`T_T4 = 40,960` elements is small: a single `4 × 128 × 768` activation is 393,216 elements, ~9.6×
over. Tensors above the threshold are ordinary rather than exceptional in real models, so the
divergence is reachable in normal use without needing any carry-over effect. **We are not claiming a
measured end-to-end training impact here; that would need its own reproducer.**

### Proposed fix

Assign the Philox counter from the **element index** rather than the thread index — i.e. derive the
per-element stream from `(seed, base_offset, element_index)` so the value at element `i` is
independent of which thread computes it, and advance the generator offset by a
**geometry-independent** amount (a function of `numel`, not of `grid.x × block.x`).

That makes the element→thread remapping disappear: the stream follows the element, so the value at
index `i` no longer depends on which thread computed it or on how many threads the device has.

*(An earlier draft justified this fix as curing "both mechanisms at once." Only one cross-device
mechanism is demonstrated here, so the rationale is stated on that one alone.)*

---

### Scope — what this report does and does not claim

**Demonstrated here:**
- **8 sampler entry points spanning 6 independent streams** all diverge above the threshold, on this
  device pair, at torch 2.11.0+cu128. *(`randn` is byte-identical to `normal_`, and `rand` to
  `uniform_`, in every row above — they share a stream. Counted honestly: **6/6 independent
  samplers**, 8/8 entry points.)*
- The transition is one element wide at 40,960 → 40,961.
- `use_deterministic_algorithms(True)` does not change the result on either device.
- The element→thread remapping in isolation, with the carried-offset path excluded by re-seeding.

**Claimed on other evidence, not on this reproducer:**
- A wider sweep in a prior campaign found **15 of 16** samplers affected. That sweep is not part of
  this report and should be treated as a separate claim with separate evidence.

**NOT claimed:**
- **Nothing about non-NVIDIA hardware.** Both devices here are NVIDIA, sharing one CUDA toolkit and
  one `libdevice`. This report is silent on cross-vendor behaviour.
- **Nothing about performance.** No timing claim is made or implied.
- **Nothing about whether this is intended behaviour.** Reproducibility across *devices* may be
  outside PyTorch's documented guarantee; the report's purpose is to establish that it is a
  *launch-geometry* property with a *predictable threshold*, and that the deterministic-algorithms
  flag does not cover it — not to assert that a guarantee was broken.
- **Two devices is n = 2** on the only parameter that matters (`SMs × maxThreadsPerSM`). The
  threshold formula predicts, and is consistent with, both — but it has not been tested against a
  third device, nor against two devices with *equal* `T` and different architecture, which is the
  measurement that would confirm the predicate as a predicate rather than as a fit to two points.

---

### Reproducer

<details><summary><code>repro_aten_offset.py</code> — 80 lines, requires only <code>torch</code></summary>

```python
# Minimal reproducer: PyTorch CUDA samplers are not reproducible across GPU
# models above a threshold set by the LAUNCH GEOMETRY, not by the seed.
#
# Run on two different NVIDIA GPUs and diff the output. Nothing but torch.
#
# MECHANISM. ATen's calc_execution_policy clamps the grid:
#     grid.x = min(SMs * (maxThreadsPerSM / 256), ceil(numel / (256*VEC)))
# and curand_init(seed, idx, offset, &state) seeds Philox with the THREAD index.
# So once numel exceeds the smaller device's thread capacity, the smaller device
# WRAPS -- one thread covers several elements -- and the element->thread mapping
# differs between the two devices. Same seed, same call, different values.
#
# THRESHOLD:  T = SMs * (maxThreadsPerSM // 256) * 256      [per device]
# Tesla T4  40 SM x 1024 thr/SM ->  40960
# NVIDIA L4 58 SM x 1536 thr/SM ->  89088
# Sizes at or below min(T) agree. The first size above it does not.
import hashlib
import torch

p = torch.cuda.get_device_properties(0)
T = p.multi_processor_count * (p.max_threads_per_multi_processor // 256) * 256
print(f"device={p.name} sm={p.multi_processor_count} thr/sm={p.max_threads_per_multi_processor} "
      f"threads_at_cap={T} torch={torch.__version__}")


def h(t):
    return hashlib.sha256(t.detach().cpu().contiguous().numpy().tobytes()).hexdigest()[:16]


def draw(fn, n, seed=1234):
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    return fn(n)


SAMPLERS = {
    "normal_":      lambda n: torch.empty(n, device="cuda").normal_(),
    "uniform_":     lambda n: torch.empty(n, device="cuda").uniform_(),
    "bernoulli_":   lambda n: torch.empty(n, device="cuda").uniform_().bernoulli_(),
    "exponential_": lambda n: torch.empty(n, device="cuda").exponential_(),
    "randn":        lambda n: torch.randn(n, device="cuda"),
    "rand":         lambda n: torch.rand(n, device="cuda"),
    "randint":      lambda n: torch.randint(0, 1 << 30, (n,), device="cuda"),
    "randperm":     lambda n: torch.randperm(n, device="cuda"),
}
# 40960 is the T4 cap. Straddle it, then straddle 4x it (the dropout family uses
# a contiguous 4-wide vector per thread, so its threshold is 4x higher).
SIZES = [40960, 40961, 163840, 163841]

print("\n--- 1. same seed, one call. Diff these lines across two GPU models. ---")
for name, fn in SAMPLERS.items():
    row = " ".join(f"{n}:{h(draw(fn, n))}" for n in SIZES)
    print(f"  {name:<13s} {row}")

print("\n--- 2. use_deterministic_algorithms(True) does NOT cover this ---")
torch.use_deterministic_algorithms(True, warn_only=True)
flagged = h(draw(SAMPLERS["normal_"], 40961))
torch.use_deterministic_algorithms(False)
unflagged = h(draw(SAMPLERS["normal_"], 40961))
print(f"  normal_(40961) flagged={flagged} unflagged={unflagged} same_on_this_device="
      f"{flagged == unflagged}   <- and BOTH differ across devices")

print("\n--- 3. isolating the remap from any carried offset ---")
# (a) within-call remap alone: RE-SEED before each call, so no offset can carry.
torch.manual_seed(7); torch.cuda.manual_seed_all(7)
a1 = torch.empty(40961, device="cuda").normal_()
torch.manual_seed(7); torch.cuda.manual_seed_all(7)
a2 = torch.empty(40961, device="cuda").normal_()
print(f"  (a) remap-only, re-seeded each call: {h(a1)} {h(a2)} equal={h(a1)==h(a2)}"
      f"   <- equal here, but the VALUE differs across devices")
# (b) offset carry-over. NOT a cross-device mechanism: these two hashes come out
#     IDENTICAL on both devices, so the offset advance did not differ with the
#     launch geometry. Retained only to show it was tested and excluded.
torch.manual_seed(7); torch.cuda.manual_seed_all(7)
_ = torch.empty(40961, device="cuda").normal_()
after_big = torch.empty(1024, device="cuda").normal_()
torch.manual_seed(7); torch.cuda.manual_seed_all(7)
alone = torch.empty(1024, device="cuda").normal_()
print(f"  (b) small draw AFTER a big one: {h(after_big)}   the same draw ALONE: {h(alone)}")
print(f"      offset carried: {h(after_big) != h(alone)}"
      f"   <- TRUE on both devices, and the two hashes MATCH across devices:"
      f" not a cross-device effect")
```

</details>

### Versions

```
torch 2.11.0+cu128
Tesla T4  (sm_75, 40 SM, 1024 max threads/SM)
NVIDIA L4 (sm_89, 58 SM, 1536 max threads/SM)
Single CUDA toolkit, single libdevice; both devices NVIDIA.
```

### Versions

```
torch 2.11.0+cu128
Tesla T4  (sm_75, 40 SM, 1024 max threads/SM)
NVIDIA L4 (sm_89, 58 SM, 1536 max threads/SM)
Single CUDA toolkit, single libdevice; both devices NVIDIA.
```

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @pbelevich @mruberry @kurtamohler

**Comment by kurtamohler** [comment]
From the first sentence of the [Reproducibility doc](https://docs.pytorch.org/docs/2.13/notes/randomness.html):

> Completely reproducible results are not guaranteed across ... different platforms.

So this does fall within intended behavior.

If the solution actually is as simple as changing a Philox counter to use an element index instead of a thread index, and if doing so doesn't cause any other issues, then perhaps it might be worth considering doing it. But there are probably plenty of other ops that will give different results on T4 vs L4, so I wouldn't really expect much benefit from it, when put in context of an entire PyTorch application.

**Comment by 59thProfile** [comment]
Thanks, agreed on the documented guarantee. I’m not arguing that PyTorch currently promises bitwise reproducibility across different GPU models, and I tried to make that explicit in the Scope section.

The narrower reason I thought this was worth reporting is that this case seems different from the usual sources of cross-device numerical variation:

1. This is RNG output itself, before any floating-point reduction, kernel-selection difference, or accumulated numerical error.
2. The divergence is not merely “T4 and L4 happen to produce different values.” It has a sharp, predictable boundary determined by launch geometry: the same seeded stream agrees through 40,960 elements and changes at exactly 40,961 for this pair.
3. The cause appears to be structural: RNG streams are attached to execution threads, so changing the number of resident threads changes which seeded stream is assigned to a given tensor element.
4. `use_deterministic_algorithms(True)` does not distinguish this case, so a user can get perfectly repeatable results on each individual GPU while silently getting a different random tensor after moving the exact same seeded workload to another NVIDIA GPU.

So I agree this is not a correctness bug relative to the current reproducibility contract. The question I’m really raising is whether CUDA RNG should unnecessarily encode launch geometry into the generated tensor when an element-indexed mapping could potentially avoid it.

I also agree the value of changing it depends heavily on cost and knock-on effects. “Use the element index” is the conceptual fix, not a claim that the implementation change is trivial or free. If preserving the current Philox consumption semantics, vectorized kernels, generator offsets, graph capture behavior, performance, etc. makes that substantially more complicated, that changes the tradeoff.

The thing I would distinguish from other T4/L4 differences is that many of those are consequences of floating-point arithmetic or deliberately different kernels. Here, for these sampling ops, the hardware dependence seems avoidable in principle because the requested operation is just “generate sample i from this seeded generator”; the GPU topology need not be part of that mapping.

There was actually an earlier issue asking for cross-device RNG consistency under the deterministic flag (#84234), so this report may be more useful as a concrete characterization of *why* that consistency fails for this family and exactly where it begins, rather than as an assertion that existing behavior violates a guarantee.

If the project considers cross-GPU RNG portability explicitly out of scope even when it can be obtained without a meaningful regression, then closing this as intended behavior makes sense. But I wanted to separate that policy decision from the mechanism, since the latter appears quite deterministic and fixable in principle.
